akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.stream.{ AbruptStageTerminationException, Attributes, FlowShape, Inlet, NeverMaterializedException, Outlet }
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.util.OptionVal
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](future: Future[Flow[In, Out, M]])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val innerMatValue = Promise[M]
+    val logic = new GraphStageLogic(shape) {
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
+          case Some(tr) =>
+            initializing.onFuture(tr)
+          case None =>
+            val cb = getAsyncCallback(initializing.onFuture)
+            future.onComplete(cb.invoke)(ExecutionContexts.parasitic)
+            //in case both ports are closed before future completion
+            setKeepGoing(true)
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!innerMatValue.isCompleted) {
+          innerMatValue.failure(new AbruptStageTerminationException(this))
+        }
+        super.postStop()
+      }
+
+      object initializing extends InHandler with OutHandler {
```

Comment:
```suggestion
      object Initializing extends InHandler with OutHandler {
```

Reply:
I think of this as a 'shortcut' for a class + val definition, so I'd prefer leaving it with the val naming notation.

Reply:
@seglo do u insist on this?

Reply:
Yes, I think we should make the change. I've never seen lowerCamelCase for object name's, other than for package objects. I'll admit I don't know if this formatting is violated elsewhere in the akka codebase, but the scala-lang.org Scala style guide says:

> Object names are like class names (upper camel case).
>
> An exception is when mimicking a package or function. This isn’t common.

https://docs.scala-lang.org/style/naming-conventions.html